lightning/backend/local.go

```diff
@@ -949,23 +950,49 @@ loopWrite:
 			for errCnt < maxRetryTimes {
 				log.L().Debug("ingest meta", zap.Reflect("meta", meta))
 				var resp *sst.IngestResponse
-				resp, err = local.Ingest(ctx, meta, region)
-				if err != nil {
-					log.L().Warn("ingest failed", log.ShortError(err), zap.Reflect("meta", meta),
-						zap.Reflect("region", region))
-					errCnt++
-					continue
-				}
+				doIngest := true
 				failpoint.Inject("FailIngestMeta", func(val failpoint.Value) {
+					// only inject the error once
 					switch val.(string) {
 					case "notleader":
-						resp.Error.NotLeader = &errorpb.NotLeader{
-							RegionId: region.Region.Id, Leader: region.Leader}
+						if errCnt == 0 {
+							resp = &sst.IngestResponse{
+								Error: &errorpb.Error{
+									NotLeader: &errorpb.NotLeader{
+										RegionId: region.Region.Id,
+										Leader:   region.Leader,
+									},
+								},
+							}
+							errCnt++
+						}
 					case "epochnotmatch":
-						resp.Error.EpochNotMatch = &errorpb.EpochNotMatch{
-							CurrentRegions: []*metapb.Region{region.Region}}
+						if retryCnt == 0 {
+							resp = &sst.IngestResponse{
+								Error: &errorpb.Error{
+									EpochNotMatch: &errorpb.EpochNotMatch{
+										CurrentRegions: []*metapb.Region{region.Region},
+									},
+								},
+							}
+							retryCnt++
+						}
+					}
+					if resp != nil {
+						err = nil
+						doIngest = false
 					}
 				})
+				if doIngest {
+					resp, err = local.Ingest(ctx, meta, region)
```

Comment:
How about inject failpoint after this line, and remove `doIngest`?

Reply:
I don't understand? the logic in this pr is to avoid actually doing Ingest since the ingest action will likely succeed.

Reply:
For some retryable error, we can directly retry the error, but a successful ingest will remove the source sst, thus the retry ingest will failed by sst file not found.